pkg/operator/ceph/object/controller.go

```diff
@@ -142,9 +144,116 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func secretPredicate() predicate.Predicate {
+	rookGV := cephv1.SchemeGroupVersion.String()
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			secret, ok := e.ObjectNew.(*corev1.Secret)
+			if !ok {
+				// not a secret
+				return false
+			}
+			// check if secret already owned by Rook:
+			for _, owner := range secret.OwnerReferences {
+				if owner.APIVersion == rookGV {
+					// already owned by Rook CR
+					return false
+				}
+			}
+			return true
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			secret, ok := e.Object.(*corev1.Secret)
+			if !ok {
+				// not a secret
+				return false
+			}
+			// check if secret already owned by Rook:
+			for _, owner := range secret.OwnerReferences {
+				if owner.APIVersion == rookGV {
+					// already owned by Rook CR
+					return false
+				}
+			}
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			secret, ok := e.Object.(*corev1.Secret)
+			if !ok {
+				// not a secret
+				return false
+			}
+			// check if secret already owned by Rook:
+			for _, owner := range secret.OwnerReferences {
+				if owner.APIVersion == rookGV {
+					// already owned by Rook CR
+					return false
+				}
+			}
+			return true
+		},
+	}
+}
+
+// Maps secret referenced by object store to the object store CR
+func mapSecretToCR(k8sClient client.Client) func(context.Context, client.Object) []reconcile.Request {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		secret, ok := obj.(*corev1.Secret)
+		if !ok {
+			return nil
+		}
+
+		// lookup object store CRs by name
+		objStores := cephv1.CephObjectStoreList{}
+		err := k8sClient.List(ctx, &objStores, client.InNamespace(secret.Namespace))
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				logger.Debugf("cephObjectStore resource for annotated secret %q not found. Ignoring since object must be deleted.", secret.Name)
+				return nil
+			}
+			logger.Errorf("failed to list cephObjectStore resources for annotated secret %q", secret.Name)
+			// Error reading the object - requeue the request.
+			return nil
+		}
+
+		var requests []reconcile.Request
+		for _, objStore := range objStores.Items {
+			found := false
+			// check if secret is referred in object store rgwConfigFromSecret:
+			for _, sec := range objStore.Spec.Gateway.RgwConfigFromSecret {
+				if sec.Name == secret.Name {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: objStore.Name, Namespace: objStore.Namespace}})
+					found = true
+					break
+				}
+			}
+			if found {
+				// secret reference found, ObjectStore added to reconcile list.
+				// check next object store
+				continue
+			}
+			if objStore.Spec.Auth.Keystone != nil && objStore.Spec.Auth.Keystone.ServiceUserSecretName == secret.Name {
+				requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: objStore.Name, Namespace: objStore.Namespace}})
+				continue
+			}
```

Comment:
I'm finding this hard to read, though I do understand what the code is doing. Maybe it would be more readable to extract the logic for determining if a CephObjectStore references a secret into a function?